File: ImgProc/tutorial_laplace_operator/Laplace_Demo.cpp
/**
 * @file Laplace_Demo.cpp
 * @brief Sample code showing how to detect edges using the Laplace operator
 * @author OpenCV team
 */

#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/imgcodecs.hpp"
#include "opencv2/highgui/highgui.hpp"
#include <stdlib.h>
#include <stdio.h>

using namespace cv;

/**
 * @function main
 */
int main( int, char** argv )
{
  //! [declare_variables]
  Mat src, src_gray, dst;
  int kernel_size = 3;
  int scale = 1;
  int delta = 0;
  int ddepth = CV_16S;
  const char* window_name = "Laplace Demo";
  //! [declare_variables]

  //! [load_image]
  /// Load an image
  src = imread( argv[1] );

  if( src.empty() )
    { return -1; }
  //! [load_image]

  //! [gaussian_blur]
  /// Remove noise by blurring with a Gaussian filter
  GaussianBlur( src, src, Size(3,3), 0, 0, BORDER_DEFAULT );
  //! [gaussian_blur]

  //! [convert_to_gray]
  /// Convert the image to grayscale
  cvtColor( src, src_gray, COLOR_RGB2GRAY );
  //! [convert_to_gray]

  /// Create window
  namedWindow( window_name, WINDOW_AUTOSIZE );

  /// Apply Laplace function
  Mat abs_dst;

  //! [laplacian]
  Laplacian( src_gray, dst, ddepth, kernel_size, scale, delta, BORDER_DEFAULT );
  //! [laplacian]
  //! [convert_scale]
  convertScaleAbs( dst, abs_dst );
  //! [convert_scale]

  //! [imshow]
  /// Show what you got
  imshow( window_name, abs_dst );
  //! [imshow]

  waitKey(0);

  return 0;
}

File: ImgProc/tutorial_laplace_operator/laplace_demo.py
import sys
import cv2

def main(argv):

    # [declare_variables]
    ddepth = cv2.CV_16S
    kernel_size = 3
    window_name = "Laplace Demo"
    # [declare_variables]

    # [load_image]
    if len(argv) < 1:
        print 'Not enough parameters'
        print 'Usage:\nlaplace_demo.py  < path_to_image >'
        return -1

    src = cv2.imread(sys.argv[1], cv2.IMREAD_COLOR)
    if src is None:
        print 'Invalid input image'
        return -1
    # [load_image]

    # [gaussian_blur]
    # Remove noise by blurring with a Gaussian filter
    src = cv2.GaussianBlur(src, (3, 3), 0)
    # [gaussian_blur]

    # [convert_to_gray]
    # Convert the image to grayscale
    src_gray = cv2.cvtColor(src, cv2.COLOR_BGR2GRAY)
    # [convert_to_gray]

    # Create Window
    cv2.namedWindow(window_name, cv2.WINDOW_AUTOSIZE)

    # [laplacian]
    # Apply Laplace function
    dst = cv2.Laplacian(src_gray, ddepth, kernel_size)
    # [laplacian]

    # [convert_scale]
    abs_dst = cv2.convertScaleAbs(dst)
    # [convert_scale]

    # [imshow]
    cv2.imshow(window_name, abs_dst)
    # [imshow]

    cv2.waitKey(0)
    return 0


if __name__ == "__main__":
    main(sys.argv[1:])

